src/Compilers/CSharp/Test/Semantic/SourceGeneration/SyntaxAwareGeneratorTests.cs

```diff
@@ -417,6 +417,172 @@ void Function()
+                    i.RegisterForSyntaxNotifications(() => receiver = new TestSyntaxReceiver());
+                    i.RegisterForPostInitialization((pic) => throw exception);
+                },
+                onExecute: (e) => { Assert.True(false); }
+                );
+
+            GeneratorDriver driver = CSharpGeneratorDriver.Create(new[] { testGenerator }, parseOptions: parseOptions);
+            driver = driver.RunGeneratorsAndUpdateCompilation(compilation, out var outputCompilation, out var outputDiagnostics);
+            var results = driver.GetRunResult();
+
+            Assert.Null(receiver);
+
+            outputDiagnostics.Verify(
+                Diagnostic("CS" + (int)ErrorCode.WRN_GeneratorFailedDuringInitialization).WithArguments("CallbackGenerator", "Exception", "test exception").WithLocation(1, 1)
+                );
+        }
+
+        [Fact]
+        public void Syntax_Receiver_Visits_Syntax_Added_In_PostInit()
+        {
+            var source = @"
+class C 
+{
+    int Property { get; set; }
+
+    void Function()
+    {
+        var x = 5;
+        x += 4;
+    }
+}
+";
+
+            var source2 = @"
+class D
+{
+    int Property { get; set; }
+
+    void Function()
+    {
+        var x = 5;
+        x += 4;
+    }
+}
+";
+            var parseOptions = TestOptions.Regular;
+            Compilation compilation = CreateCompilation(source, options: TestOptions.DebugDll, parseOptions: parseOptions);
+            compilation.VerifyDiagnostics();
+
+            Assert.Single(compilation.SyntaxTrees);
+
+            ISyntaxReceiver? receiver = null;
+
+            var testGenerator = new CallbackGenerator(
+                onInit: (i) =>
+                {
+                    i.RegisterForSyntaxNotifications(() => new TestSyntaxReceiver());
+                    i.RegisterForPostInitialization((pic) => pic.AddSource("postInit", source2));
+                },
+                onExecute: (e) => receiver = e.SyntaxReceiver
+                );
+
+            GeneratorDriver driver = CSharpGeneratorDriver.Create(new[] { testGenerator }, parseOptions: parseOptions);
+            driver.RunGeneratorsAndUpdateCompilation(compilation, out _, out _);
+
+            Assert.NotNull(receiver);
+            Assert.IsType<TestSyntaxReceiver>(receiver);
+
+            TestSyntaxReceiver testReceiver = (TestSyntaxReceiver)receiver!;
+            Assert.Equal(42, testReceiver.VisitedNodes.Count);
+            Assert.IsType<CompilationUnitSyntax>(testReceiver.VisitedNodes[0]);
+            Assert.IsType<ClassDeclarationSyntax>(testReceiver.VisitedNodes[1]);
+            Assert.Equal("C", ((ClassDeclarationSyntax)testReceiver.VisitedNodes[1]).Identifier.Text);
+            Assert.IsType<CompilationUnitSyntax>(testReceiver.VisitedNodes[21]);
+            Assert.IsType<ClassDeclarationSyntax>(testReceiver.VisitedNodes[22]);
+            Assert.Equal("D", ((ClassDeclarationSyntax)testReceiver.VisitedNodes[22]).Identifier.Text);
```

Comment:
Consider avoiding specific counts and indices. Perhaps just:
```C#
var classDeclarations = testReceiver.VisitedNodes.OfType<ClassDeclarationSyntax>().Select(c => c.Identifier.Text);
Assert.Equal(new[] { "C", "D" }, classDeclarations);
```

Same comment for test below.